{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "99659970",
   "metadata": {},
   "source": [
    "# 转录因子活性推断\n",
    "\n",
    "使用 SCENIC 进行转录因子活性推断"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "80374dcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "392491ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "import exprmat as em\n",
    "\n",
    "# set working directory\n",
    "em.setwd('/home/data/yangz/shared/tutorial/single-cell-rna')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c3c48cde",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "   ━━━━━━━━━━━━━━━━━━━━━━━━━━ loading samples          4 / 4     (00:05 < 00:00)\n"
     ]
    }
   ],
   "source": [
    "expm = em.load_experiment('expression', load_samples = True, load_subset = 'mono-neutro')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "767fc273",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mannotated data\u001b[0m of size 10160 × 19389\n",
      "\u001b[31msubset\u001b[0m mono-neutro of size 10160 × 19389\n",
      "contains modalities: \u001b[36mrna\u001b[0m\n",
      "\n",
      " modality \u001b[36m[rna]\u001b[0m\n",
      "\u001b[32m    obs\u001b[0m : sample \u001b[90;3m<cat>\u001b[0m batch \u001b[90;3m<cat>\u001b[0m group \u001b[90;3m<cat>\u001b[0m modality \u001b[90;3m<cat>\u001b[0m taxa \u001b[90;3m<cat>\u001b[0m barcode \u001b[90;3m<o>\u001b[0m ubc \u001b[90;3m<o>\u001b[0m\n",
      "          n.umi \u001b[90;3m<f64>\u001b[0m n.genes \u001b[90;3m<i64>\u001b[0m n.mito \u001b[90;3m<f64>\u001b[0m n.ribo \u001b[90;3m<f64>\u001b[0m pct.mito \u001b[90;3m<f64>\u001b[0m pct.ribo \u001b[90;3m<f64>\u001b[0m\n",
      "          filter \u001b[90;3m<bool>\u001b[0m score.doublet \u001b[90;3m<f64>\u001b[0m score.doublet.se \u001b[90;3m<f64>\u001b[0m is.doublet \u001b[90;3m<bool>\u001b[0m qc \u001b[90;3m<bool>\u001b[0m\n",
      "          leiden \u001b[90;3m<cat>\u001b[0m cell.type \u001b[90;3m<cat>\u001b[0m score \u001b[90;3m<f64>\u001b[0m potency \u001b[90;3m<cat>\u001b[0m relative \u001b[90;3m<f64>\u001b[0m score.preknn \u001b[90;3m<f64>\u001b[0m\n",
      "          potency.preknn \u001b[90;3m<cat>\u001b[0m ppt.pseudotime \u001b[90;3m<f64>\u001b[0m ppt.seg \u001b[90;3m<cat>\u001b[0m ppt.edge \u001b[90;3m<cat>\u001b[0m\n",
      "          ppt.milestones \u001b[90;3m<cat>\u001b[0m\n",
      "\u001b[32m    var\u001b[0m : chr \u001b[90;3m<cat>\u001b[0m start \u001b[90;3m<i64>\u001b[0m end \u001b[90;3m<i64>\u001b[0m strand \u001b[90;3m<cat>\u001b[0m id \u001b[90;3m<o>\u001b[0m subtype \u001b[90;3m<cat>\u001b[0m gene \u001b[90;3m<cat>\u001b[0m tlen \u001b[90;3m<f64>\u001b[0m\n",
      "          cdslen \u001b[90;3m<i64>\u001b[0m assembly \u001b[90;3m<cat>\u001b[0m uid \u001b[90;3m<o>\u001b[0m vst.hvg \u001b[90;3m<bool>\u001b[0m vst.all.means \u001b[90;3m<f64>\u001b[0m vst.all.vars \u001b[90;3m<f64>\u001b[0m\n",
      "          vst.all.vars.norm \u001b[90;3m<f64>\u001b[0m vst.all.hvg.rank \u001b[90;3m<f32>\u001b[0m vst.all.hvg \u001b[90;3m<bool>\u001b[0m n.umi \u001b[90;3m<f64>\u001b[0m symbol \u001b[90;3m<cat>\u001b[0m\n",
      "\u001b[32m layers\u001b[0m : counts \u001b[90;3m<f32>\u001b[0m magic \u001b[90;3m<f64>\u001b[0m norm \u001b[90;3m<f32>\u001b[0m\n",
      "\u001b[32m   obsm\u001b[0m : cnmf.10 \u001b[90;3m<df>\u001b[0m diffmap \u001b[90;3m<arr:f32(5)>\u001b[0m harmony \u001b[90;3m<arr:f64(35)>\u001b[0m knn \u001b[90;3m<arr:i32(100)>\u001b[0m\n",
      "          knn.d \u001b[90;3m<arr:f32(100)>\u001b[0m knn.d.nn.scvi \u001b[90;3m<arr:f32(50)>\u001b[0m knn.d.nn30.diffmap \u001b[90;3m<arr:f32(30)>\u001b[0m\n",
      "          knn.nn.scvi \u001b[90;3m<arr:i32(50)>\u001b[0m knn.nn30.diffmap \u001b[90;3m<arr:i32(30)>\u001b[0m pca \u001b[90;3m<arr:f64(35)>\u001b[0m\n",
      "          ppt \u001b[90;3m<arr:f64(2000)>\u001b[0m scvi \u001b[90;3m<arr:f32(30)>\u001b[0m umap \u001b[90;3m<arr:f32(2)>\u001b[0m umap.diff \u001b[90;3m<arr:f32(2)>\u001b[0m\n",
      "\u001b[32m   varm\u001b[0m : cnmf.10 \u001b[90;3m<arr:f64(10)>\u001b[0m cnmf.coef.10 \u001b[90;3m<arr:f64(10)>\u001b[0m pca \u001b[90;3m<arr:f64(35)>\u001b[0m\n",
      "\u001b[32m   obsp\u001b[0m : connectivities \u001b[90;3m<csr:f32>\u001b[0m connectivities.nn.scvi \u001b[90;3m<csr:f32>\u001b[0m\n",
      "          connectivities.nn30.diffmap \u001b[90;3m<csr:f32>\u001b[0m distances \u001b[90;3m<csr:f32>\u001b[0m distances.nn.scvi \u001b[90;3m<csr:f32>\u001b[0m\n",
      "          distances.nn30.diffmap \u001b[90;3m<csr:f32>\u001b[0m\n",
      "\u001b[32m    uns\u001b[0m : cell.type.colors cell.type_colors cnmf cnmf.args cnmf.density.10 cnmf.dist.10 cnmf.stats\n",
      "          diffmap leiden leiden.colors lr lr.hsa magic.errors magic.t markers neighbors nn.scvi\n",
      "          nn30.diffmap pca ppt ppt.graph ppt.milestones.colors ppt.pseudotime ppt.seg.colors scvi\n",
      "          trace umap umap.diff\n",
      "\n",
      "\u001b[31m[*]\u001b[0m composed of samples:\n",
      "  distal     \u001b[36mrna  \u001b[0m \u001b[33mmmu \u001b[0m   batch \u001b[32mb1     \u001b[0m   of size \u001b[32m6599\u001b[0m × \u001b[33m21689\u001b[0m\n",
      "  niche      \u001b[36mrna  \u001b[0m \u001b[33mmmu \u001b[0m   batch \u001b[32mb2     \u001b[0m   of size \u001b[32m3908\u001b[0m × \u001b[33m21970\u001b[0m\n",
      "  normal     \u001b[36mrna  \u001b[0m \u001b[33mmmu \u001b[0m   batch \u001b[32mb3     \u001b[0m   of size \u001b[32m4627\u001b[0m × \u001b[33m21917\u001b[0m\n",
      "  metacell   \u001b[36mrna  \u001b[0m \u001b[33mmmu \u001b[0m   batch \u001b[32mautogen\u001b[0m   of size \u001b[32m1384\u001b[0m × \u001b[33m19078\u001b[0m\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(expm)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c663e0d",
   "metadata": {},
   "source": [
    "### SCENIC 分析\n",
    "\n",
    "为了减小计算量并增大稳定性，我们在元细胞上计算 SCENIC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "45b71182",
   "metadata": {},
   "outputs": [],
   "source": [
    "expm.modalities['rna']['metacell'].layers['counts'] = em.np.expm1(\n",
    "    expm.modalities['rna']['metacell'].X\n",
    ").floor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "93b2b0d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[36m[i] \u001b[0mloaded 1860 transcriptional factors.\n",
      "\u001b[0m\u001b[36m[i] \u001b[0mstarting grnboost2 using 80 processes\n",
      "\u001b[0m"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "   ━━━━━━━━━━━━━━━━━━━ inferring partial network 19078 / 19078 (1:10:46 < 00:00)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[36m[i] \u001b[0mfinished in 4249.272406101227 seconds.\n",
      "\u001b[0m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[36m[i] \u001b[0mcalculating pearson correlations ...\n",
      "\u001b[0m\u001b[36m[i] \u001b[0musing 100 workers.\n",
      "\u001b[0m\u001b[36m[i] \u001b[0mcreating regulons from a dataframe of enriched features.\n",
      "\u001b[0m\u001b[36m[i] \u001b[0madditional columns saved: []\n",
      "\u001b[0m"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "   ━━━━━━━━━━━━━━━━━━━━━━━━━━ auc                    236 / 236   (00:05 < 00:00)\n"
     ]
    }
   ],
   "source": [
    "expm.run_rna_infer_tf_activity(\n",
    "    run_on_samples = ['metacell'],\n",
    "\n",
    "    layer = 'counts',\n",
    "    gene = 'gene',\n",
    "    method = 'grnboost2',\n",
    "    taxa = 'mmu',\n",
    "    ncpus = 100,\n",
    "    seed = 42,\n",
    "\n",
    "    features = 'genes',\n",
    "    cistromes = 'motifs',\n",
    "    thresholds = [0.75, 0.90],\n",
    "    top_n_targets = [50],\n",
    "    top_n_regulators = [5, 10, 50],\n",
    "    min_genes = 20,\n",
    "    mask_dropout = True,\n",
    "    keep_only_activating = True,\n",
    "    rank_threshold = 5000,\n",
    "    auc_threshold = 0.05,\n",
    "    nes_threshold = 3.0,\n",
    "    max_similarity_fdr = 0.001,\n",
    "    min_orthologous_identity = 0,\n",
    "    chunk_size = 100,\n",
    "\n",
    "    weighted = False,\n",
    "\n",
    "    key_adjacencies = 'tf.adjacencies',\n",
    "    key_added = 'tf'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1ad086fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tf</th>\n",
       "      <th>target</th>\n",
       "      <th>importance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>450</th>\n",
       "      <td>Anxa1</td>\n",
       "      <td>S100a8</td>\n",
       "      <td>2.318543e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>450</th>\n",
       "      <td>Anxa1</td>\n",
       "      <td>S100a9</td>\n",
       "      <td>2.195228e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442</th>\n",
       "      <td>Prdx5</td>\n",
       "      <td>S100a8</td>\n",
       "      <td>2.173934e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442</th>\n",
       "      <td>Prdx5</td>\n",
       "      <td>S100a9</td>\n",
       "      <td>2.107148e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>500</th>\n",
       "      <td>Zeb2</td>\n",
       "      <td>S100a8</td>\n",
       "      <td>1.612040e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>769</th>\n",
       "      <td>Abcf2</td>\n",
       "      <td>Rsph9</td>\n",
       "      <td>2.961198e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>Xrcc4</td>\n",
       "      <td>Hbb-bt</td>\n",
       "      <td>2.957419e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>627</th>\n",
       "      <td>Dhx36</td>\n",
       "      <td>Has1</td>\n",
       "      <td>4.824099e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1218</th>\n",
       "      <td>Pkm</td>\n",
       "      <td>Has1</td>\n",
       "      <td>1.101952e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>930</th>\n",
       "      <td>Foxo3</td>\n",
       "      <td>Rsph9</td>\n",
       "      <td>2.376227e-21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>929062 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         tf  target    importance\n",
       "450   Anxa1  S100a8  2.318543e+01\n",
       "450   Anxa1  S100a9  2.195228e+01\n",
       "442   Prdx5  S100a8  2.173934e+01\n",
       "442   Prdx5  S100a9  2.107148e+01\n",
       "500    Zeb2  S100a8  1.612040e+01\n",
       "...     ...     ...           ...\n",
       "769   Abcf2   Rsph9  2.961198e-19\n",
       "145   Xrcc4  Hbb-bt  2.957419e-19\n",
       "627   Dhx36    Has1  4.824099e-20\n",
       "1218    Pkm    Has1  1.101952e-20\n",
       "930   Foxo3   Rsph9  2.376227e-21\n",
       "\n",
       "[929062 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expm.modalities['rna']['metacell'].uns['tf.adjacencies']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "05d6cd5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>regulon</th>\n",
       "      <th>Ahdc1 (+)</th>\n",
       "      <th>Ahr (+)</th>\n",
       "      <th>Arid3a (+)</th>\n",
       "      <th>Arnt (+)</th>\n",
       "      <th>Atf1 (+)</th>\n",
       "      <th>Atf3 (+)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>mc:integrated:1</th>\n",
       "      <td>0.091195</td>\n",
       "      <td>0.039358</td>\n",
       "      <td>0.027912</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.050292</td>\n",
       "      <td>0.068105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mc:integrated:2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.038827</td>\n",
       "      <td>0.034006</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004394</td>\n",
       "      <td>0.109103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mc:integrated:3</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.048634</td>\n",
       "      <td>0.045927</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.034814</td>\n",
       "      <td>0.069670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mc:integrated:4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.040191</td>\n",
       "      <td>0.036249</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.030777</td>\n",
       "      <td>0.045177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mc:integrated:5</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.020117</td>\n",
       "      <td>0.039077</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.010616</td>\n",
       "      <td>0.043169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mc:integrated:1380</th>\n",
       "      <td>0.008386</td>\n",
       "      <td>0.050358</td>\n",
       "      <td>0.026132</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.062848</td>\n",
       "      <td>0.102932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mc:integrated:1381</th>\n",
       "      <td>0.068284</td>\n",
       "      <td>0.025631</td>\n",
       "      <td>0.029009</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011330</td>\n",
       "      <td>0.099359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mc:integrated:1382</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.040364</td>\n",
       "      <td>0.028375</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015188</td>\n",
       "      <td>0.096717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mc:integrated:1383</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.045030</td>\n",
       "      <td>0.037443</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.014987</td>\n",
       "      <td>0.037382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mc:integrated:1384</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.041914</td>\n",
       "      <td>0.011360</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.010148</td>\n",
       "      <td>0.141997</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1384 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "regulon             Ahdc1 (+)   Ahr (+)  Arid3a (+)  Arnt (+)  Atf1 (+)  \\\n",
       "mc:integrated:1      0.091195  0.039358    0.027912       0.0  0.050292   \n",
       "mc:integrated:2      0.000000  0.038827    0.034006       0.0  0.004394   \n",
       "mc:integrated:3      0.000000  0.048634    0.045927       0.0  0.034814   \n",
       "mc:integrated:4      0.000000  0.040191    0.036249       0.0  0.030777   \n",
       "mc:integrated:5      0.000000  0.020117    0.039077       0.0  0.010616   \n",
       "...                       ...       ...         ...       ...       ...   \n",
       "mc:integrated:1380   0.008386  0.050358    0.026132       0.0  0.062848   \n",
       "mc:integrated:1381   0.068284  0.025631    0.029009       0.0  0.011330   \n",
       "mc:integrated:1382   0.000000  0.040364    0.028375       0.0  0.015188   \n",
       "mc:integrated:1383   0.000000  0.045030    0.037443       0.0  0.014987   \n",
       "mc:integrated:1384   0.000000  0.041914    0.011360       0.0  0.010148   \n",
       "\n",
       "regulon             Atf3 (+)  \n",
       "mc:integrated:1     0.068105  \n",
       "mc:integrated:2     0.109103  \n",
       "mc:integrated:3     0.069670  \n",
       "mc:integrated:4     0.045177  \n",
       "mc:integrated:5     0.043169  \n",
       "...                      ...  \n",
       "mc:integrated:1380  0.102932  \n",
       "mc:integrated:1381  0.099359  \n",
       "mc:integrated:1382  0.096717  \n",
       "mc:integrated:1383  0.037382  \n",
       "mc:integrated:1384  0.141997  \n",
       "\n",
       "[1384 rows x 6 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expm.modalities['rna']['metacell'].obsm['tf'].iloc[:, 0:6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "754b4ad1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[36m[i] \u001b[0mresident memory: 4.60 GiB\n",
      "\u001b[0m\u001b[36m[i] \u001b[0mvirtual memory: 20.18 GiB\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "em.memory()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "40c977d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[36m[i] \u001b[0mmain dataset write to expression/subsets/mono-neutro.h5mu\n",
      "\u001b[0m\u001b[36m[i] \u001b[0msaving individual samples. (pass `save_samples = False` to skip)\n",
      "\u001b[0m"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "   ━━━━━━━━━━━━━━━━━━━━━━━━━━ modality [rna]           4 / 4     (00:02 < 00:00)\n"
     ]
    }
   ],
   "source": [
    "expm.save()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "exprmat",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
